### PR TITLE
Fixes #33801 - Replace obsolete URI.escape method

### PR DESCRIPTION
**Pull request by adiabramovitch**

### CODE REVIEW

File: lib/proxy_api/external_ipam.rb [hunk 1]
```diff
@@ -63,7 +63,7 @@ def initialize(args)
     #     {"error": "Unable to connect to External IPAM server"}
     def next_ip(subnet, mac, group = "")
       raise "subnet cannot be nil" if subnet.nil?
-      response = parse get("/subnet/#{subnet}/next_ip?mac=#{mac}&group=#{URI.escape(group.to_s)}")
+      response = parse get("/subnet/#{subnet}/next_ip?mac=#{mac}&group=#{Addressable::URI.encode(group.to_s)}")
```
Comment (ekohl) [inline lib/proxy_api/external_ipam.rb]: We don't depend on Addressable. At least not directly. This means there's no guarantee this gem is available.

I wonder if using `URI.encode_www_form` is correct. I think it would be something like this:
```suggestion
      response = parse get("/subnet/#{subnet}/next_ip?#{URI.encode_www_form(mac: mac, group: group.to_s)}")
```

Though I'm not 100% sure it is correct.

Comment (lzap) [reply lib/proxy_api/external_ipam.rb]: Using CGI.escape is the correct way of doing this now.

Comment (adiabramovitch) [reply lib/proxy_api/external_ipam.rb]: I tried both of them in the resource.rb file and the output is different from what is expected.

Comment (ekohl) [reply lib/proxy_api/external_ipam.rb]: Could you share the examples you tried?

Comment (adiabramovitch) [reply lib/proxy_api/external_ipam.rb:66]: I wasn't sure how to use the URI.encode_www_form because I have a different path at each time. 

This is what I checked: 
image

And these are the errors:
 
* when using CGI.escape: 
image

* when using  URI.encode_www_form_component:
image

Comment (adiabramovitch) [reply lib/proxy_api/external_ipam.rb]: Maybe I am missing something but I got errors when using these methods and I wasn't sure how to use the URI.encode_www_form because there is a different path at each time.

This is what I checked:
image
 
* when using CGI.escape:
image

* when using  URI.encode_www_form_component:
image

Comment (ekohl) [reply lib/proxy_api/external_ipam.rb]: For future reference: please copy-paste it rather than a screenshot. Consuming plain text is much easier than screenshots.

For your paths you should only encode the components, not the entire path.

Edit: When I mean components, I'm referring to how [RFC3986](https://datatracker.ietf.org/doc/html/rfc3986#section-3) describes the parts. In this API both path and query but I'm talking about [the RFC3986 path section](https://datatracker.ietf.org/doc/html/rfc3986#section-3.3) and [RFC3986 query section](https://datatracker.ietf.org/doc/html/rfc3986#section-3.4).

Comment (Ron-Lavi) [reply lib/proxy_api/external_ipam.rb]: Katello had the same issue and they fixed it by switching to `URI.encode_www_form` as @ekohl suggested, see [Katello - Fixes #32995 - URI.escape is obsolete warning](https://github.com/Katello/katello/pull/9477/commits/de337537b505f9a4447d230eb43205034223cf94#diff-6ff2fabe1c0b1475dcf25fedcb7546b9bc0336efc40e3c6443a2b2b43d0bd576R154)

**Comment by lzap** [comment]
I think this is the problem:

```ruby
>> require "uri"
=> false
>> require "cgi"
=> false

>> URI.escape("http://example.com/?a=\11\15")
(irb):9: warning: URI.escape is obsolete
=> "http://example.com/?a=%09%0D"

>> CGI.escape("http://example.com/?a=\11\15")
=> "http%3A%2F%2Fexample.com%2F%3Fa%3D%09%0D"
```

I think you will have to break parameters up and escape each separately.

File: lib/proxy_api/resource.rb [hunk 2]
```diff
@@ -64,14 +64,15 @@ def parse(response)
...
 
     # Perform GET operation on the supplied path
-    def get(path = nil, payload = {})
+    def get(path = nil, additional_headers = {})
       with_logger do
         telemetry_duration_histogram(:proxy_api_duration, :ms, method: 'get') do
           # This ensures that an extra "/" is not generated
           if path
-            resource[URI.escape(path)].get payload
+            path = "?#{URI.encode_www_form(CGI.parse(URI.parse(path).query))}" unless URI.parse(path).query.nil?
```
Comment (Ron-Lavi) [inline lib/proxy_api/resource.rb]: wouldn't it return only the query?

didn't test it yet, though how about something like:
```ruby
    def escape_path(path)
      parsed_path = URI.parse(path)
      query = parsed_path.query
      path = "#{parsed_path.path}?#{CGI.escape(query)}" unless query.nil?
      path
    end

    # Perform GET operation on the supplied path
    def get(path = nil, additional_headers = {})
      with_logger do
        telemetry_duration_histogram(:proxy_api_duration, :ms, method: 'get') do
          # This ensures that an extra "/" is not generated
          if path
            resource[escape_path(path)].get additional_headers
          else
            resource.get additional_headers
          end
        end
      end
    end
```

Comment (adiabramovitch) [reply lib/proxy_api/resource.rb]: I think you are right, I update the PR with your suggestion, Thanks!

**Review by Ron-Lavi — DISMISSED** [review]
code wise LGTM 👍🏼
@lzap @ekohl could you take another look?

File: lib/proxy_api/resource.rb [hunk 3]
```diff
@@ -63,15 +63,22 @@ def parse(response)
...
+    def escape_path(path)
+      parsed_path = URI.parse(path)
+      query = parsed_path.query
+      path = "#{parsed_path.path}?#{CGI.escape(query)}" unless query.nil?
+      path
+    end
+
     # Perform GET operation on the supplied path
-    def get(path = nil, payload = {})
+    def get(path = nil, additional_headers = {})
```
Comment (tbrisker) [inline lib/proxy_api/resource.rb]: why are we changing the parameter name here?

File: lib/proxy_api/external_ipam.rb [hunk 4]
```diff
@@ -63,7 +63,7 @@ def initialize(args)
     #     {"error": "Unable to connect to External IPAM server"}
     def next_ip(subnet, mac, group = "")
       raise "subnet cannot be nil" if subnet.nil?
-      response = parse get("/subnet/#{subnet}/next_ip?mac=#{mac}&group=#{URI.escape(group.to_s)}")
+      response = parse get("/subnet/#{subnet}/next_ip?mac=#{mac}&group=#{CGI.escape(group.to_s)}")
```
Comment (tbrisker) [inline lib/proxy_api/external_ipam.rb]: How about:
```suggestion
      response = parse get("/subnet/#{subnet}/next_ip?#{URI.encode_www_form(mac: mac, group: group)}")
```

File: lib/proxy_api/resource.rb [hunk 5]
```diff
@@ -63,15 +63,22 @@ def parse(response)
...
       with_logger do
         telemetry_duration_histogram(:proxy_api_duration, :ms, method: 'get') do
           # This ensures that an extra "/" is not generated
```
Comment (tbrisker) [inline lib/proxy_api/resource.rb:78]: This comment seems to indicate this escaping might not what we actually want to do here?

File: lib/proxy_api/resource.rb [hunk 6]
```diff
@@ -63,15 +63,22 @@ def parse(response)
       false
     end
 
+    def escape_path(path)
+      parsed_path = URI.parse(path)
+      query = parsed_path.query
+      path = "#{parsed_path.path}?#{CGI.escape(query)}" unless query.nil?
```
Comment (tbrisker) [inline lib/proxy_api/resource.rb]: Are you sure this will escape it properly? for example:
```
[82] pry(main)> a= URI.parse('/aa?page=5')
=> #<URI::Generic /aa?page=5>
[83] pry(main)> CGI.escape(a.query)
=> "page%3D5"
[84] pry(main)> URI.escape '/aa?page=5'
=> "/aa?page=5"

**Review by tbrisker — CHANGES_REQUESTED** [review]

File: lib/proxy_api/resource.rb (hunk 3 above)
Comment (adiabramovitch) [reply lib/proxy_api/resource.rb]: Ewoud ask to change the name

File: lib/proxy_api/resource.rb (hunk 6 above)
Comment (adiabramovitch) [reply lib/proxy_api/resource.rb]: this is what the output will be: "/aa?page%3D5"
after adding the parsed_path.path part.

Comment (tbrisker) [reply lib/proxy_api/resource.rb]: which is incorrect for query parameters, you don't want to escape the = sign

File: lib/proxy_api/external_ipam.rb (hunk 4 above)
Comment (adiabramovitch) [reply lib/proxy_api/external_ipam.rb]: yes, you are right, thank you/
Do you think I should change the other references? they have only one param from what I saw.

File: lib/proxy_api/resource.rb (hunk 3 above)
Comment (ekohl) [reply lib/proxy_api/resource.rb]: If you look at https://rubydoc.info/gems/rest-client/RestClient/Resource#get-instance_method you'll see that it's not really a payload (which would generally be the body) but additional headers. GET never has a body so this makes it match the API it actually calls.

File: lib/proxy_api/resource.rb [hunk 7]
```diff
@@ -63,15 +63,22 @@ def parse(response)
       false
     end
 
+    def escape_path(path)
+      parsed_path = URI.parse(path)
+      query = parsed_path.query
+      path = "#{parsed_path.path}?#{URI.encode_www_form(CGI.parse(query))}" unless query.nil?
+      path
+    end
```
Comment (ekohl) [inline lib/proxy_api/resource.rb]: This doesn't look correct to me. While I don't expect a problem with correct input, it doesn't help anything. With incorrect input you actually get a very different output.

For example:
```
[1] pry(main)> path = URI.parse('http://example.com/?parameter1=a text with & in it')
=> #<URI::HTTP http://example.com/?parameter1=a%20text%20with%20&%20in%20it>
[2] pry(main)> path.query
=> "parameter1=a%20text%20with%20&%20in%20it"
[3] pry(main)> require 'cgi'
=> true
[4] pry(main)> CGI.parse(path.query)
=> {"parameter1"=>["a text with "], " in it"=>[]}
[5] pry(main)> URI.encode_www_form(CGI.parse(path.query))
=> "parameter1=a+text+with+&"
```

Now this is garbage input, but that makes me think we should stop escaping path and requiring the caller to properly escape. That's the only place where it can really be done reliably.

**Review by ekohl — DISMISSED** [review]
On IRC I recommended this as an approach:
https://github.com/ekohl/foreman/commit/20e009d657281be50d269e0807f36cd35d552bd4
It makes it reliable to detect if the new API is called or a legacy one.

IMHO long term `get()` should not escape at all and trust the caller to do the right thing. Doing escaping in multiple places is asking for problems.

**Comment by lzap** [comment]
Yes that looks sane, I would suggest also to update all proxy API implementations to take usage of that @adiabramovitch.

**Comment by adiabramovitch** [comment]
@ekohl, @lzap  Thank you for the clarification. I will do that change

**Review by Ron-Lavi — DISMISSED** [review]
Code wise LGTM,
@lzap @ekohl @tbrisker can you take another look?

File: lib/proxy_api/resource.rb [hunk 8]
```diff
@@ -64,14 +64,20 @@ def parse(response)
     end
 
     # Perform GET operation on the supplied path
-    def get(path = nil, payload = {})
+    def get(path = nil, query: nil, headers: {})
```
Comment (tbrisker) [inline lib/proxy_api/resource.rb]: This change in the method signature will break all consumers of this function in plugins as well.
You could maintain the existing signature using something like:
```suggestion
    def get(path = nil, payload = {})
      query = payload.delete(:query)
```

However, looking at the [RestClient API](https://www.rubydoc.info/gems/rest-client/1.6.1/RestClient/Resource#get-instance_method) and history of this method, I think this additional parameter was never actually needed or used.
The only resource I found which uses it is the BMC resource which passes all of its args as the second parameter. Looking at the [smart proxy side](https://github.com/theforeman/smart-proxy/blob/develop/modules/bmc/bmc_api.rb), AFAICT there is nothing there that looks at any additional headers - so these additional headers were simply ignored.
It also makes sense that in the unlikely case a specific resource does require setting additional request headers, those would be set at the resource level, not at the api call stage.

To sum it up - I would clean up the BMC controller so it doesn't set headers, and add a deprecation here so we could switch it to using a keyword argument instead:
```suggestion
    def get(path = nil, payload = {})
      query = payload.delete(:query)
      Foreman::Deprecation.deprecation_warning("3.3", "passing additional headers to ProxyApi resource GET action") unless payload.empty?
```
Then in 3.3 we can drop the warning and change the method signature to
```suggestion
    def get(path = nil, query: nil)
```

**Review by tbrisker — CHANGES_REQUESTED** [review]

File: lib/proxy_api/resource.rb (hunk 8 above)
Comment (adiabramovitch) [reply lib/proxy_api/resource.rb:67]: @tbrisker thanks :)

**Comment by tbrisker** [comment]
@adiabramovitch can you rebase on latest develop please so katello tests pass?

**Review by tbrisker — APPROVED** [review]
Thanks @adiabramovitch !
